Advanced Mode

Jacky720 opened the pull request:
Another fairly big one! It's only fair I write a full changelog.
1. **Add "Advanced Mode", allowing modification (mostly with dropdowns and checkboxes) of all 513 flags in the file.** (Flag edits automatically reflect back and forth with the normal interface, e.g. changing flag 5 alters the fun value above.)
2. Remove the "unknown kills" value, as it's reset upon entry into battle and therefore purposeless in the save file. (It's flag 401 and helps track the difference between before and after a battle to properly increment global.kills).
3. Hide the cell phone options when you don't have a cell phone, and add a toggle for the "Papyrus and Undyne" phone (which otherwise shares its ID with the normal Papyrus's Phone). Resolves #8.
4. Consolidate `floweyStates`, `torielStates`, `dogamyDogaressaStates`, etc, into one manageable `stateChoiceArrays` array directly linking each set of options to the ID of its dropdown.
5. ***Massively* expand plot value options, now containing every value that can theoretically be obtained in a save file.** Unobtainable values are currently commented in the list.
6. Merge `loadSelectFromObj` into `updateSelection`, and always rewrite options. New parameters for `updateSelection` are frequently entirely optional (it directly accesses `stateChoiceArrays` instead of always requiring a list of options to use), allowing it to be called in a number of useful ways.
7. Add simple annotations to the massive `start` function.

Jacky720 committed:
Update to gh-pages + cell-slot hiding

Hide cell phone slots if the phone is not obtained, to avoid confusion.

Jacky720 committed:
Full-file modification, Pap+Undyne phone, and more

I appear to have lost track of everything I changed here. Whoops.
- Add an "advanced mode" to modify every flag in the file. Still looking into linking it with the basic mode to prevent confusion about which value will actually be saved.
- Add a toggle for flag 465 that corresponds with the phone names.
- General cleanup (no, really, I forgot)

Jacky720 committed:
Merge branch 'patch-1' into patch-2

Jacky720 committed:
Merge pull request #2 from Jacky720/patch-2

Resolve "behind stream" problems. Hopefully.

Jacky720 committed:
Remove unkkills (it's redundant), add/rm newlines

Jacky720 committed:
Persist advanced mode through reload

"I thought I was the only one with that power... But apparently YOUR desires for this world override MINE."
- Flowey, upon learning that cookies exist.

Jacky720 committed:
Include flag number, shorten some flag names

Jacky720 committed:
Link form inputs to flag inputs

Heck, this could cut down saving to almost entirely use the flags field. It doesn't. But it could.

Jacky720 committed:
Merge pull request #3 from Jacky720/patch-1

Advanced Mode!

Jacky720 committed:
BIG object reduction

Convert option arrays into one dictionary of option arrays with input IDs as keys. Begin simplifying other stuff using this (loadSelectFromObject, your days are numbered).

Jacky720 committed:
Correct Mettaton state flag

important bugfix

Jacky720 committed:
"false" is truthy, patch

boo(lean)

Jacky720 committed:
Eliminate loadSelectFromObj, refactor updateSelection

Jacky720 committed:
Add default cases to bugged tables

Previous workaround prevented *ever* adding a 0 value to these arrays. Remove that.

Jacky720 committed:
Add precise plot values through Undyne

Going to need a way to filter these later.

Jacky720 committed:
Complete plot value list

aw yeaaaaaaah

Jacky720 committed:
Merge branch 'gh-pages' into patch-1

Jacky720 committed:
Merge pull request #4 from Jacky720/patch-1

Plot values, default cases, rm redundant funcs

Jacky720 committed:
rm initializations; updateSelection does 'em anyway

this will probably break

Jacky720 committed:
...use more updateSelection

Say, what would happen if you previously had such an incredibly invalid location? Looks like it would just break. Not anymore!

Jacky720 committed:
Workaround for Flowey phase dropdown

I feel like it's either this or make an "Empty file" preset that it loads before loading this one... which would probably have the same effect.

Jacky720 committed:
[m] comment unobtainable plot value

Now that I've gotten it to not throw a fit about not having a zero case...

Jacky720 committed:
Sort decimal keys (plot values) correctly

I know I shouldn't iterate over the array a second time, but... ugh, do I need to make another array with the "decimals" values for each stateChoiceArray?

Jacky720 committed:
rm semicolon

That's what I get for blind copy-paste.

Jacky720 committed:
Add descriptions + options to flags.js

Now to make use of it.

Jacky720 committed:
Make basicBool not an array

I should be able to sort by typeof flags[n][2] now

Jacky720 committed:
Attempt to utilise new flags.js

Fingers crossed!

Jacky720 committed:
Bugfixing #1

Jacky720 committed:
Correct checkbox functionality

Jacky720 committed:
Add explanation as tooltip

Jacky720 committed:
Remove undefined tooltips

Jacky720 committed:
Clean Advanced section header

Jacky720 committed:
Simplify advanced mode toggle, re-sync checkboxes

Jacky720 committed:
Move advanced tags around

Jacky720 committed:
[m] Fix pcy link

Jacky720 committed:
Move explanation out of section

To prevent it from being hidden

Jacky720 committed:
[m] Fix margin

Jacky720 committed:
Fix advanced toggle button appearance

Jacky720 committed:
Fix checkboxes always checking

why is "0" truthy again

Jacky720 committed:
[m] Fix checkboxes never checking

Jacky720 committed:
[m] Use _blank links

Jacky720 committed:
Set value directly for inputs

Previously set the attribute, like options. Hopefully fix flag 512 initialization.

Jacky720 committed:
Debugging

Jacky720 committed:
More debugging

Jacky720 committed:
It turns out "debug" is taken

Jacky720 committed:
Correct var name

again

Jacky720 committed:
Initialize debugging globally

...this problem is weirding me out.

Jacky720 committed:
Print value?

how it feels to be toby fox

Jacky720 committed:
Continued debugging

what the fun is wrong

Jacky720 committed:
Try newField.value

Jacky720 committed:
Test hunch.

Jacky720 committed:
More debugging

Jacky720 committed:
Continue to add debug messages

I'm getting closer.

Jacky720 committed:
Flag 512 is not saved.

Hooray! Bug fixed!

Jacky720 committed:
Remove debugging code

Jacky720 committed:
Add some color classes

Jacky720 committed:
Add some flag highlighting

Jacky720 committed:
Further darken grayed-out text

Jacky720 committed:
Merge branch 'crumblingstatue:gh-pages' into patch-1

Jacky720 committed:
Merge pull request #5 from Jacky720/gh-pages

update patch branch

Jacky720 committed:
Update checkboxes on load

Jacky720 committed:
Correct Undyne's Letter EX value

why is there no value 2...

Jacky720 committed:
Correct True Lab key flags

Curse you, zero-indexing! Except for every other flag in the game! But for this!

Jacky720 committed:
Use updateSelection with flags

Jacky720 committed:
Add flags 310 and 322

Not that they do anything...

KockaAdmiralac commented on main.js:
```diff
@@ -769,17 +893,56 @@ function updateIniFromForm(ini) {
+    
+    // Sanitize value
+    if (typeof value === "string") {
+        value = parseInt(value.trim());
```
```suggestion
        value = Number(value.trim());
```
Probably in other places where `parseInt` is being used, too, unless you're using it to parse non-decimal numbers.

KockaAdmiralac commented on main.js:
```diff
@@ -769,17 +893,56 @@ function updateIniFromForm(ini) {
+    if (typeof value === "string") {
+        value = parseInt(value.trim());
+    } else if (typeof value === "number") {
+        value = parseInt(value);
```
If the value is already a number, why do we need to parse it?

KockaAdmiralac commented on main.js:
```diff
@@ -769,17 +893,56 @@ function updateIniFromForm(ini) {
+        value = parseInt(value.trim());
+    } else if (typeof value === "number") {
+        value = parseInt(value);
+    } else if (value == undefined) {
+        if (select.value == undefined) {
```
```suggestion
    } else if (value === undefined) {
        if (select.value === undefined) {
```

KockaAdmiralac commented on main.js:
```diff
@@ -769,17 +893,56 @@ function updateIniFromForm(ini) {
+    }
+    
+    // Create options
+    for (var key of Object.keys(stateChoiceArrays[id]).sort((a, b) => a - b)) { // (Decimal keys don't automatically sort correctly)
```
Do you have Internet Explorer nearby to test whether Flowey's Time Machine works correctly on it? Because I would rather not drop support for it if it was already supported.

Jacky720 replied:
It took a reboot, but now I do. And... it doesn't work. `display: grid` seems the obvious problem, but, more importantly, the scripts aren't running. I don't know IE's dev tools well enough to find a fix for that if an easy one exists.

Jacky720 replied:
Okay, looks like the string literal at line 982 is the issue. Not sure if there are more, but combined with the total display failure I think we can safely say IE is no longer supported.

KockaAdmiralac commented on main.js:
```diff
@@ -860,7 +1038,7 @@ function updateSaveValuesFromForm(values) {
     values[97] = document.getElementById("sav-papyrusstate").value;
     values[111] = document.getElementById("sav-shyrenstate").value;
-    values[231] = document.getElementById("sav-unkkills").value;
+    //values[231] = document.getElementById("sav-unkkills").value;
```
?

Jacky720 replied:
sav-unkkills no longer exists, per point 2.

KockaAdmiralac replied:
I read the description after I posted the review for some reason, but shouldn't it be removed completely then?

Jacky720 replied:
That's fair.

KockaAdmiralac commented on main.js:
```diff
@@ -871,6 +1049,7 @@ function updateSaveValuesFromForm(values) {
     values[427] = document.getElementById("sav-muffetstate").value;
     values[432] = document.getElementById("sav-broguardsstate").value;
     values[455] = document.getElementById("sav-mettatonstate").value;
+    values[495] = +document.getElementById("sav-undyne-cell").checked;
```
This should probably be more explicit (in other places, too):
```suggestion
    values[495] = Number(document.getElementById("sav-undyne-cell").checked);
```

KockaAdmiralac commented on style.css:
```diff
@@ -247,3 +254,15 @@ th {
+  color: #ff0000;
+}
+
+.hidden {
+    display: none;
+}
```
Now that we have this, can you apply it to the `type="file"` element from `index.html` and remove the rule about hiding these from above?

KockaAdmiralac reviewed (changes requested):
Mostly requesting comments on old browser compatibility.

Jacky720 committed:
Hide file options

Because they're ugly and this lets us kill some style code

Jacky720 committed:
Remove redundant display: none

Well, now it's redundant.

Jacky720 committed:
Merge pull request #6 from Jacky720/patch-1

Minor fixes

Jacky720 committed:
Use Number()

Jacky720 committed:
Add commas

(whoops)

Jacky720 committed:
Remove sav-unkkills

It was as if it was never there at all.

Jacky720 committed:
Prevent true lab checkbox overriding advanced mode

Previously, this checkbox would have set the value to 0 otherwise, wiping out Alphys's date if it had occurred in the save previously. That's bad.

Jacky720 committed:
[m] Better explain flag 8

KockaAdmiralac commented on index.html:
```diff
@@ -493,6 +487,18 @@ <h1>SAVE Data (file0)</h1>
+            <div class="col-span-2">
+              <p>PLEASE only proceed if you know what you are doing!</p>
+              <p>Each flag has a mouseover explanation.</p>
+              <p><a href="https://pcy.ulyssis.be/undertale/flags" target="_blank">More</a> <a href="https://docs.google.com/document/d/1d5dt7A_jkTBhduVEGbWGtDbLEYMXCxSLUvqIqeXtYKw/" target="_blank">info</a></p>
```
Thoughts on having this extracted into the help page alongside an explanation of the interface?

Jacky720 replied:
I'm sure there's more useful information to add to the help page on that, but my thoughts are "that's a good idea" and "done".

KockaAdmiralac reviewed (approved):
Minor comment, let me know when you think this is ready for merging.

Jacky720 committed:
Add geno removal tips

I *do* have information on this. Not all the information; but I have information.

Jacky720 committed:
Advanced description and some key flags

Jacky720 committed:
close tag

Maybe it won't email me about messing up this time?

Jacky720 committed:
...correct tags

Heh, yeah, I totally know Unordered Lists and List Items apart from each other, heh!

Jacky720 committed:
Link help.html

Jacky720 committed:
Un-bold lists

(why does this need to use local style...)

Jacky720 committed:
subpage link

Jacky720 committed:
Add ID for linking
